File: packages/apple/Sources/Composable/Guardrail.swift
import Foundation

// MARK: - Guardrail Result

/// Result of a guardrail check
public enum GuardrailResult: Sendable {
    case passed
    case blocked(reason: String)
    case modified(newText: String, reason: String)
}

// MARK: - Guardrail Protocol

/// Protocol for input/output validation and safety.
///
/// ```swift
/// let guardrail = InputLengthGuardrail(maxCharacters: 4000)
/// let result = try await guardrail.checkInput(input)
/// ```
public protocol Guardrail: Sendable {
    var name: String { get }

    /// Check input before it reaches the model
    func checkInput(_ input: ChainInput) async throws -> GuardrailResult

    /// Check output before it is returned to the caller
    func checkOutput(_ output: ChainOutput) async throws -> GuardrailResult
}

// MARK: - Input Length Guardrail

/// Validates input length against model constraints
public struct InputLengthGuardrail: Guardrail {
    public let name = "InputLengthGuardrail"
    private let maxCharacters: Int
    private let truncate: Bool

    public init(maxCharacters: Int = 16000, truncate: Bool = true) {
        self.maxCharacters = maxCharacters
        self.truncate = truncate
    }

    public func checkInput(_ input: ChainInput) async throws -> GuardrailResult {
        if input.text.count <= maxCharacters { return .passed }
        if truncate {
            return .modified(
                newText: String(input.text.prefix(maxCharacters)),
                reason: "Input truncated from \(input.text.count) to \(maxCharacters) characters"
            )
        }
        return .blocked(reason: "Input exceeds maximum length of \(maxCharacters) characters")
    }

    public func checkOutput(_ output: ChainOutput) async throws -> GuardrailResult { .passed }
}

// MARK: - Content Filter Guardrail

/// Blocks or filters sensitive content patterns
public struct ContentFilterGuardrail: Guardrail {
    public let name = "ContentFilterGuardrail"
    private let blockedPatterns: [String]

    public init(blockedPatterns: [String] = []) {
        self.blockedPatterns = blockedPatterns
    }

    public func checkInput(_ input: ChainInput) async throws -> GuardrailResult {
        for pattern in blockedPatterns {
            if input.text.localizedCaseInsensitiveContains(pattern) {
                return .blocked(reason: "Input contains blocked content pattern")
            }
        }
        return .passed
    }

    public func checkOutput(_ output: ChainOutput) async throws -> GuardrailResult {
        for pattern in blockedPatterns {
            if output.text.localizedCaseInsensitiveContains(pattern) {
                return .blocked(reason: "Output contains blocked content pattern")
            }
        }
        return .passed
    }
}

// MARK: - Guarded Chain

/// Wraps a chain with guardrail checks on input and output
@available(iOS 15.0, macOS 14.0, *)
public struct GuardedChain: Chain {
    public let name: String
    private let chain: any Chain
    private let guardrails: [any Guardrail]

    public init(name: String? = nil, chain: any Chain, guardrails: [any Guardrail]) {
        self.name = name ?? "Guarded(\(chain.name))"
        self.chain = chain
        self.guardrails = guardrails
    }

    public func invoke(_ input: ChainInput) async throws -> ChainOutput {
        var currentInput = input

        // Check input guardrails
        for guardrail in guardrails {
            switch try await guardrail.checkInput(currentInput) {
            case .passed:
                continue
            case .blocked(let reason):
                throw LocanaraError.invalidInput("Blocked by \(guardrail.name): \(reason)")
            case .modified(let newText, _):
                currentInput = ChainInput(text: newText, metadata: currentInput.metadata)
            }
        }

        // Execute chain
        let output = try await chain.invoke(currentInput)

        // Check output guardrails
        for guardrail in guardrails {
            switch try await guardrail.checkOutput(output) {
            case .passed:
                continue
            case .blocked(let reason):
                throw LocanaraError.executionFailed("Output blocked by \(guardrail.name): \(reason)")
            case .modified(let newText, _):
                return ChainOutput(
                    value: newText,
                    text: newText,
                    metadata: output.metadata,
                    processingTimeMs: output.processingTimeMs
                )
            }
        }

        return output
    }
}

File: packages/android/locanara/src/main/kotlin/com/locanara/composable/Guardrail.kt
package com.locanara.composable

import com.locanara.core.ChainInput
import com.locanara.core.ChainOutput

/**
 * Result of a guardrail check
 */
sealed class GuardrailResult {
    data object Passed : GuardrailResult()
    data class Blocked(val reason: String) : GuardrailResult()
    data class Modified(val newText: String, val reason: String) : GuardrailResult()
}

/**
 * Input/output validation and safety interface
 */
interface Guardrail {
    val name: String

    /** Check input before it reaches the model */
    suspend fun checkInput(input: ChainInput): GuardrailResult

    /** Check output before it is returned to the caller */
    suspend fun checkOutput(output: ChainOutput): GuardrailResult
}

/**
 * Validates input length against model constraints
 */
class InputLengthGuardrail(
    private val maxCharacters: Int = 16000,
    private val truncate: Boolean = true
) : Guardrail {
    override val name = "InputLengthGuardrail"

    override suspend fun checkInput(input: ChainInput): GuardrailResult {
        if (input.text.length <= maxCharacters) return GuardrailResult.Passed
        return if (truncate) {
            GuardrailResult.Modified(
                input.text.take(maxCharacters),
                "Input truncated from ${input.text.length} to $maxCharacters characters"
            )
        } else {
            GuardrailResult.Blocked("Input exceeds maximum length of $maxCharacters characters")
        }
    }

    override suspend fun checkOutput(output: ChainOutput) = GuardrailResult.Passed
}

/**
 * Blocks or filters sensitive content patterns
 */
class ContentFilterGuardrail(
    private val blockedPatterns: List<String> = emptyList()
) : Guardrail {
    override val name = "ContentFilterGuardrail"

    override suspend fun checkInput(input: ChainInput): GuardrailResult {
        for (pattern in blockedPatterns) {
            if (input.text.contains(pattern, ignoreCase = true)) {
                return GuardrailResult.Blocked("Input contains blocked content")
            }
        }
        return GuardrailResult.Passed
    }

    override suspend fun checkOutput(output: ChainOutput): GuardrailResult {
        for (pattern in blockedPatterns) {
            if (output.text.contains(pattern, ignoreCase = true)) {
                return GuardrailResult.Blocked("Output contains blocked content")
            }
        }
        return GuardrailResult.Passed
    }
}

/**
 * Wraps a chain with guardrail checks on input and output
 */
class GuardedChain(
    private val chain: Chain,
    private val guardrails: List<Guardrail>,
    override val name: String = "Guarded(${chain.name})"
) : Chain {
    override suspend fun invoke(input: ChainInput): ChainOutput {
        var currentInput = input

        for (guardrail in guardrails) {
            when (val result = guardrail.checkInput(currentInput)) {
                is GuardrailResult.Passed -> continue
                is GuardrailResult.Blocked ->
                    throw IllegalArgumentException("Blocked by ${guardrail.name}: ${result.reason}")
                is GuardrailResult.Modified ->
                    currentInput = ChainInput(result.newText, currentInput.metadata)
            }
        }

        val output = chain.invoke(currentInput)

        for (guardrail in guardrails) {
            when (val result = guardrail.checkOutput(output)) {
                is GuardrailResult.Passed -> continue
                is GuardrailResult.Blocked ->
                    throw IllegalStateException("Output blocked by ${guardrail.name}: ${result.reason}")
                is GuardrailResult.Modified ->
                    return ChainOutput(result.newText, result.newText, output.metadata, output.processingTimeMs)
            }
        }

        return output
    }
}
